ADR-180 (#181)
decentraland/adr: content/ADR-25-explorer-repositories-decoupling-.md

We are planning to support desktop clients along with the web implementation, and currently our codebase is tightly coupled to the browsers.

When aiming for different platforms we face new development streamlines and independent release lifecycles. Also, code duplication should be minimized.

> With those constraints in mind, this document proposes the initial steps towards a project reorganization that fulfill those needs.

Reviewer: ```suggestion
With those constraints in mind, this document proposes the initial steps toward a project reorganization that fulfills those needs.
```

The CI/CD of this repository is going to output internal and production releases for play.decentraland.zone and play.decentraland.org respectively.

- Contains React and the entry point that instantiates Renderer and Kernel.

> - The typescript project will also be responsible for creating the ethereum provider and passing it as a parameter to Kernel.

Reviewer: ```suggestion
- The typescript project will also be responsible for creating the Ethereum provider and passing it as a parameter to Kernel.
```

> - The typescript project contains a thin layer with the `unity-interface` and all the code that is directly related with the browser renderer (GIFs, mouse events wrapper, audio, video, etc).

Reviewer: ```suggestion
- The typescript project contains a thin layer with the `unity-interface` and all the code that is directly related to the browser renderer (GIFs, mouse events wrapper, audio, video, etc).
```

> - An independent Unity project contains the Renderer Package, the web specific entry point (InitialScene + Main + etc) and any web specific components that may be needed.

Reviewer: ```suggestion
- An independent Unity project contains the Renderer Package, the web-specific entry point (InitialScene + Main + etc), and any web-specific components that may be needed.
```

![resources/ADR-25-explorer-repositories-decoupling/image-2.svg](resources/ADR-25-explorer-repositories-decoupling/image-2.svg)

### Desktop Shell Repository

Contains all the components to run a desktop version of Decentraland.

The CI/CD of this repository is going to be bound to internal and production releases that will be served from a link in the download section of our landing pages (TBD).

> - The Unity executable will be run by a native JS virtual machine (i.e. NodeJS) to run the scenes and some parts of the kernel. This will be done via a shell command, and this means two processes will be running by this stage. The JS project will communicate with the Unity process using IPC (Exact definition TBD) or FFI linking. In the future there will be different executables generated for each of the desktop platforms (win, mac, linux).

Reviewer: ```suggestion
- The Unity executable will be run by a native JS virtual machine (i.e. NodeJS) to run the scenes and some parts of the kernel. This will be done via a shell command, which means two processes will be running at this stage. The JS project will communicate with the Unity process using IPC (Exact definition TBD) or FFI linking. In the future, different executables will be generated for each desktop platform (win, mac, Linux).
```

> - A Desktop Unity project containing the renderer’s package and specific platform components (i.e. sign in/sign up screens, terms of service, desktop ethereum provider, any communication glue for the Native JS Project)

Reviewer: ```suggestion
- A Desktop Unity project containing the renderer’s package and specific platform components (i.e. sign-in/sign-up screens, terms of service, desktop Ethereum provider, and any communication glue for the Native JS Project)
```

> - On the Native JS Project, a thin TS layer with the `unity-interface` desktop implementation and all the code that is directly related with the desktop renderer will be included. This implementation will have the specifics of the IPC communication needed.

Reviewer: ```suggestion
- On the Native JS Project, a thin TS layer with the `unity-interface` desktop implementation and all the code that is directly related to the desktop renderer will be included. This implementation will have the specifics of the IPC communication needed.
```

![resources/ADR-25-explorer-repositories-decoupling/image-3.svg](resources/ADR-25-explorer-repositories-decoupling/image-3.svg)

## Staged plan

- Separate a copy of the explorer repo into a decentraland/renderer repository. Keep Explorer working.

> - In isolation, work on the decentraland/renderer repository to make CI work and expose interfaces to be used by kernel in the future.

Reviewer: ```suggestion
- In isolation, work on the decentraland/renderer repository to make CI work and expose interfaces to be used by the kernel in the future.
```

> - After a checkpoint (TBD) start using the new build in decentraland/explorer synchronize changes and delete unity files from the original repository.

Reviewer: ```suggestion
- After a checkpoint (TBD) start using the new build in decentraland/explorer, synchronize changes and delete unity files from the original repository.
```

> - After stable release, repeat the same process to separate website and kernel.

Reviewer: ```suggestion
- After stable release, repeat the same process to separate the website and kernel.
```

## Benefit

> This approach decouples the kernel repository from any business logic tied to the web browsers, moving that part directly to the web shell repository. It not only enables better testing tools (NodeJS) but it also makes the code more portable, following the objective of reusing parts of the `kernel` in the desktop distribution.

Reviewer: ```suggestion
This approach decouples the kernel repository from any business logic tied to the web browsers, moving that part directly to the web shell repository. It not only enables better testing tools (NodeJS) but also makes the code more portable, following the objective of reusing parts of the `kernel` in the desktop distribution.
```
